### PR TITLE
tree: Add and use isTreeNode

### PR DESCRIPTION
**Pull request by CraigMacomber**
## Description

Since TreeNodes are a bit strange, and mess with their prototypes as well as have an Unhydrated state, a special isTreeNode function has need added and used instead of instanceof which is documented as not being usable for them.

This is extracted from https://github.com/microsoft/FluidFramework/pull/19834 (which was updated to include tests and clarify handling of unhydrated content before this PR was made).

## Reviewer Guidance

The review process is outlined on [this wiki page](https://github.com/microsoft/FluidFramework/wiki/PR-Guidelines#guidelines).

### CODE REVIEW

File: packages/dds/tree/src/simple-tree/proxies.ts [hunk 1]
```diff
@@ -47,6 +47,24 @@ import { tryGetFlexNodeTarget, setFlexNode, getFlexNode, tryGetFlexNode } from "
 import { cursorFromFieldData, cursorFromNodeData } from "./toMapTree.js";
 import { RawTreeNode, createRawNode, extractRawNodeContent } from "./rawNode.js";
 
+/**
+ * Detects if the given 'candidate' is a TreeNode.
```
Comment (Josmithr) [inline packages/dds/tree/src/simple-tree/proxies.ts:51]: Great docs! Thanks!!

File: packages/dds/tree/src/simple-tree/proxies.ts [hunk 2]
```diff
@@ -47,6 +47,24 @@ import { tryGetFlexNodeTarget, setFlexNode, getFlexNode, tryGetFlexNode } from "
...
+ *
+ * Because the common usage is to check if a value being inserted/set is a TreeNode,
+ * this function permits calling with primitives as well as objects.
+ *
+ * Primitives will always return false (as they are copies of data, not references to nodes).
+ *
+ * @param candidate - Value which may be a TreeNode
+ * @returns true if the given 'candidate' is a hydrated TreeNode.
+ */
+export function isTreeNode(candidate: unknown): candidate is TreeNode | Unhydrated<TreeNode> {
```
Comment (Josmithr) [inline packages/dds/tree/src/simple-tree/proxies.ts:64]: We don't expose `TreeNode`s as a part of the public API, right? So this has no impact on user workflows?

File: packages/dds/tree/src/test/simple-tree/proxies.spec.ts [hunk 3]
```diff
@@ -6,9 +6,12 @@
 import { strict as assert } from "assert";
 import { MockHandle } from "@fluidframework/test-runtime-utils";
 import { TreeArrayNode, NodeFromSchema, SchemaFactory } from "../../simple-tree/index.js";
+// TODO: test other things from "proxies" file.
```
Comment (Josmithr) [inline packages/dds/tree/src/test/simple-tree/proxies.spec.ts:9]: Do we have an item filed for this?

**Review by Josmithr — APPROVED** [review]
🚀

File: packages/dds/tree/src/simple-tree/proxies.ts (hunk 2 above)
Comment (Josmithr) [reply packages/dds/tree/src/simple-tree/proxies.ts:64]: I guess we don't expect users to need to make this sort of check. That's really more what I wanted to verify.

File: packages/dds/tree/src/test/simple-tree/proxies.spec.ts (hunk 3 above)
Comment (CraigMacomber) [reply packages/dds/tree/src/test/simple-tree/proxies.spec.ts:9]: Most of our 500+ TODOs in this package don't have items. I think ones like this that are tracking a minor code quality issue and not a bug arn't worth littering our public codebase with internal work item references, and can't be tracked on github as stalebot would close them before we fixed it in most cases.

File: packages/dds/tree/src/simple-tree/proxies.ts (hunk 2 above)
Comment (CraigMacomber) [reply packages/dds/tree/src/simple-tree/proxies.ts:64]: We might want/need to export it at some point, but not for now.

**Pull request merged**